packages/user_packages/iotcc_mqtt.py

```diff
@@ -76,50 +99,73 @@ def run(self, registry):
+            while reg_attempts <= retry_attempts:
                 try:
-                    # Register edge system (gateway)
-                    self.iotcc.set_properties(self.iotcc_edge_system, {"key1": "value1", "key2": "value2"})
+                    self.iotcc_edge_system = self.iotcc.register(edge_system)
                     break
-                except Exception:
-                    # In the third attempt if get exception raise it
-                    if attempts == max_retry_attempts:
+                except Exception, e:
+                    if reg_attempts == retry_attempts:
                         raise
-                    attempts += 1
-                    # The sleep time before re-trying depends on the infrastructure requirement of broker to restart
-                    # It can be modified or removed as per the infrastructure requirement
-                    time.sleep(5)
+                    reg_attempts += 1
+                    log.error('MQTT Connection Failed - {0}'.format(str(e)))
+                    log.info('Trying Edge System {0} Registration: Attempt - {1}'.format(edge_system.name,
+                                                                                         str(reg_attempts)))
+                    time.sleep(delay_retries)
+
             registry.register("iotcc_mqtt", self.iotcc)
-            # Store the registered edge system object in liota package manager registry after the
             registry.register("iotcc_mqtt_edge_system", self.iotcc_edge_system)
 
-        except RegistrationFailure:
-            print "EdgeSystem registration to IOTCC failed"
+            # Attempts for setting edge system properties
+            prop_attempts = 0
+            # Set Properties with retry attempts in case of exception
+            while prop_attempts < retry_attempts:
+                try:
+                    self.iotcc.set_properties(self.iotcc_edge_system,
+                                              {"Country": "USA-G", "State": "California", "City": "Palo Alto",
+                                               "Location": "VMware HQ", "Building": "Promontory H Lab",
+                                               "Floor": "First Floor"})
+                    break
+                except Exception, e:
+                    prop_attempts += 1
+                    log.error(
+                        'Exception while setting Property for Edge System {0} - {1}'.format(edge_system.name, str(e)))
+                    log.info('Trying setting properties for Edge System {0}: Attempt - {1}'.format(edge_system.name,
+                                                                                                   str(prop_attempts)))
+                    time.sleep(delay_retries)
+
+        except Exception:
+            log.error("EdgeSystem registration to IOTCC failed even after all the retries starting cleanup process")
+            self.clean_up()
```

Comment:
If a gateway restarts after a few days, and tries properties registration and no response is received from Pulse, do we want to send unregister() edge system call to Pulse, deleting the whole history?

Reply:
The property call doesn't raise the exception after multiple attempt failure, hence cleanup of the entire history is not done.

But agreed we should remove the unregister call in case of multiple retry exception and should keep it only in the clean_up method with comments. I`ll fix it in both the packages.